internal/registry/registry.go

```diff
@@ -571,12 +604,116 @@ func (r *Registry) loadPersistedProviders(ctx context.Context) (int, error) {
 			return 0, err
 		}
 
+		if pinfo.Publisher != nil && pinfo.Publisher.ID == pinfo.AddrInfo.ID && equalAddrs(pinfo.Publisher.Addrs, pinfo.AddrInfo.Addrs) {
+			pinfo.Publisher = &pinfo.AddrInfo
+		}
+
 		r.providers[peerID] = pinfo
 		count++
 	}
 	return count, nil
 }
 
+func (r *Registry) convertV0ToV1(ctx context.Context, value []byte) error {
+	type v0ProviderInfo struct {
+		AddrInfo              peer.AddrInfo
+		DiscoveryAddr         string    `json:",omitempty"`
+		LastAdvertisement     cid.Cid   `json:",omitempty"`
+		LastAdvertisementTime time.Time `json:",omitempty"`
+		Publisher             peer.ID   `json:",omitempty"`
+	}
+
+	v0Info := new(v0ProviderInfo)
+	err := json.Unmarshal(value, v0Info)
+	if err != nil {
+		return fmt.Errorf("cannot unmarshal v0 provider data: %w", err)
+	}
+
+	v1Info := ProviderInfo{
+		AddrInfo:              v0Info.AddrInfo,
+		DiscoveryAddr:         v0Info.DiscoveryAddr,
+		LastAdvertisement:     v0Info.LastAdvertisement,
+		LastAdvertisementTime: v0Info.LastAdvertisementTime,
+	}
+
+	if v0Info.Publisher == v1Info.AddrInfo.ID {
+		v1Info.Publisher = &v1Info.AddrInfo
+	} else if v0Info.Publisher.Validate() == nil {
+		v1Info.Publisher = &peer.AddrInfo{
+			ID: v0Info.Publisher,
+		}
+	}
+
+	value, err = json.Marshal(v1Info)
+	if err != nil {
+		return fmt.Errorf("cannot marshal v1 provider data: %w", err)
+	}
+
+	if err = r.dstore.Put(ctx, v1Info.dsKey(), value); err != nil {
+		return fmt.Errorf("could not write v1 provider data: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Registry) migrate(ctx context.Context) error {
```

Comment:
nit. It would be nice if this and the other migration related funcs where in a separate file.

Reply:
The PR description should highlight that this change can not be easily reverted because it does DB migration.

The PR description should also include steps to revert this change.

Reply:
will move migrate to a separate file/utility and provide a revert function.